### ISSUE TITLE
IPv6 support changes broke region servers listed by hostname

### ISSUE THREAD

**Issue by mrosile** [SUSE-Enceladus/cloud-regionsrv-client#59]
Since the [addition of IPv6 support](https://github.com/SUSE-Enceladus/cloud-regionsrv-client/commit/1162e9553b271e604c1b0d4f30f07f99d4b3f0a0) back in April, the "regionsrv" value in /etc/regionserverclnt.cfg  seems to expect either IPv4 or IPv6 addresses only -- preventing region servers to be listed using host names.

Is this expected?

If expected, do the SSL/TLS certificates have to then be updated to also reference the IP address(es)?

**Comment by rjschwei** [comment]
> Since the [addition of IPv6 support](https://github.com/SUSE-Enceladus/cloud-regionsrv-client/commit/1162e9553b271e604c1b0d4f30f07f99d4b3f0a0) back in April, the "regionsrv" value in /etc/regionserverclnt.cfg seems to expect either IPv4 or IPv6 addresses only -- preventing region servers to be listed using host names.
> 
> Is this expected?

No the configured value is expected to be a mix of hostnames and IPv4 or IPv6 addresses in any order or mixture the user sees fit.

> 
> If expected, do the SSL/TLS certificates have to then be updated to also reference the IP address(es)?

The certs on the server have to reflect whatever is configured. If the server is configured in `/etc/regionserverclnt.cfg` with the name then the server cert must have that name in it, for example.

**Comment by rjschwei** [comment]
@bear454 can you take a look at this please?

**Comment by bear454** [comment]
@mrosile can you please post the trace and the log for where you're having an error? I'm unable to reproduce an error when using FQDN instead of IP address.

**Comment by mrosile** [comment]
Here's the error returned by /usr/sbin/registercloudguest:
`Traceback (most recent call last):
  File "/usr/sbin/registercloudguest", line 279, in <module>
    region_smt_data = utils.fetch_smt_data(cfg, proxies)
  File "/usr/lib/python3.6/site-packages/cloudregister/registerutils.py", line 248, in fetch_smt_data
    ip_addr = ipaddress.ip_address(srv.strip())
  File "/usr/lib64/python3.6/ipaddress.py", line 54, in ip_address
    address)
ValueError: 'na**********p.mcp-services.net' does not appear to be an IPv4 or IPv6 address`

This is from a fresh install of SLES 15 SP3, with client, cloud-regionsrv-client-9.2.1-6.51.1.noarch.

This is the relevant [server] section of /etc/regionserverclnt.cfg:

`[server]
api = regionInfo
certLocation = /var/lib/regionService/certs
regionsrv = na**********p.mcp-services.net,eu**********p.mcp-services.net,au**********p.mcp-services.net,ap**********p.mcp-services.net,af**********p.mcp-services.net`

Hostnames were purposely obfuscated with asterisks (*).

**Comment by rjschwei** [comment]
@mrosile can you test the changes in #60 Thanks, and sorry for missing this.

**Comment by mrosile** [comment]
@rjschwei The current changes introduce new problems -- I added comments to #60 .

**Comment by rjschwei** [comment]
@mrosile stupid mistake, sorry about that. Fixed.

**Comment by mrosile** [comment]
@rjschwei 
One list item to resolve ... around line 274, this:
`region_servers += random.shuffle(region_servers_dns)`
should be:
`region_servers.append( random.shuffle(region_servers_dns) )`

Note the replacement of `+=` with `.append`

**Comment by mrosile** [comment]
@rjschwei 
Two things:
First, around line 264 where the `ip_addr` is being compared to see if its an IPv6 address, I think the same should be done with the `else` statement, that appends to the IPv4 address list, but rather than using `else`, I think this should be an "else if" (`elif`).  Otherwise, hostname records will be appended to the IPv4 address list, which is not desirable.

Resultant code:
```python
            if isinstance(ip_addr, ipaddress.IPv6Address):
                region_servers_ipv6.append(ip_addr)
            elif isinstance(ip_addr, ipaddress.IPv4Address):
                region_servers_ipv4.append(ip_addr)
```
Second, around line 274, the use of `random.shuffle` does not return any value, shuffle only modifies the original list.  Attempting to append to the `region_servers` list fails in this case due to a "NoneType" attempting to be appended.

Similar to above lines calling `random.shuffle` on the IP address lists, the same should be done here for `region_servers_dns`; call random.shuffle first, then append to the `region_servers` list.

Resultant code:
```python
random.shuffle(region_servers_dns)
region_servers += region_servers_dns
```

**Comment by rjschwei** [comment]
Thanks for the review. Problem was a missing `continue` in the exception. The exception case identifies the specified server as a DNS name as it is not an IP address. After processing the name in the exception clause we need to pick up the next entry from the list rather than proceeding into the code the processes addressed based entries. Fixed now in #60

**Comment by mrosile** [comment]
@rjschwei The second item, regarding `random.shuffle` remains outstanding.

**Comment by rjschwei** [comment]
Alright, learned something new

```
>>> import random
>>> a = [1,2,3,4,5]
>>> random.shuffle(a)
>>> a
[3, 5, 2, 1, 4]
>>> b = []
>>> b += random.shuffle(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object is not iterable
```
Fixed in #60. Thanks for your patience.

**Comment by mrosile** [comment]
Thank you @rjschwei for the quick turn around, I can confirm these changes are working!
